### PR TITLE
Add random_state

### LINKED ISSUE #139: Add option to seed random number generator
In #89, we added a random seed to our examples. I think it would be nice if we had an option to set the seed globally, e.g. `ws.set_seed` for the OO API.

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #139.

So far, MVARICA supports the `random_state` argument. I'll add the other functions soon. In the meantime, please let me know if my changes are OK @mbillingr.

### CODE REVIEW

**Comment by cle1109** [comment]
I don't understand the coverage report:

1. How come the report is shown _before_ the Travis build for Python 3.5 is finished? I thought this build triggers Coveralls?
2. Why did coverage increase? What is the reference/baseline? The previous commit? The current master?

**Comment by cle1109** [comment]
@mbillingr can you take a look at https://github.com/cle1109/scot/commit/c83eb5b4f05445a8ea018a79c4973948355658c5 - I've disabled two tests in this function because I have no idea what they do. Also the last line in the test raised an error related to the eigenvalue decomposition (see Travis log in the previous commit) - I think not only is the input data linearly dependent(?), but there is also no `assert` statement to catch any errors.

**Comment by cle1109** [comment]
Either I'm completely blind or there really is a problem with `conda`. Since I've tried long enough to resolve this, I reported the issue at https://github.com/ContinuumIO/anaconda-issues/issues/683.

**Comment by cle1109** [comment]
I guess that's all - do you see anything I forgot?

**Comment by cle1109** [comment]
Furthermore, what's up with Travis?

File: scot/csp.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def csp(x, cl, numcomp=None):
         sigma2 += np.cov(x2[t, :, :]) / x2.shape[0]
     sigma2 /= sigma2.trace()
 
-    e, w = eig(sigma1, sigma1 + sigma2, overwrite_a=True, overwrite_b=True, check_finite=False)
+    e, w = eigh(sigma1, sigma1 + sigma2, overwrite_a=True, overwrite_b=True, check_finite=False)
```
Comment (mbillingr) [inline scot/csp.py]: How is this change related to the random state?

Comment (cle1109) [inline scot/csp.py]: It's not :-). But when I stumbled over it, I thought it had to be changed.

File: scot/datatools.py [hunk 2]
```diff
@@ -154,17 +168,48 @@ def randomize_phase(data):
...
     """
+    rng = check_random_state(random_state)
     data = np.asarray(data)
     data_freq = np.fft.rfft(data)
-    data_freq = np.abs(data_freq) * np.exp(1j*np.random.random_sample(data_freq.shape)*2*np.pi)
+    data_freq = np.abs(data_freq) * np.exp(1j*rng.random_sample(data_freq.shape)*2*np.pi)
     return np.fft.irfft(data_freq, data.shape[-1])
 
 
-def atleast_3d(x):
```
Comment (mbillingr) [inline scot/datatools.py]: Why did you remove this function - and why are things still working after removing it?

Comment (cle1109) [inline scot/datatools.py]: I didn't remove it, I just moved it to the top (which isn't necessary of course, but in the course of debugging the cyclic import I tried moving it around).

File: scot/tests/test_ooapi.py [hunk 3]
```diff
@@ -238,6 +238,25 @@ def test_plotting(self):
...
+        api.do_mvarica(random_state=1)
+        mixing2 = api.mixing_
+        assert_allclose(mixing1, mixing2)
+
+        # test CSPVARICA
+        api.do_cspvarica(random_state=1)
+        mixing1 = api.mixing_
+        api.do_cspvarica(random_state=1)
+        mixing2 = api.mixing_
+        assert_allclose(mixing1, mixing2)
```
Comment (mbillingr) [inline scot/tests/test_ooapi.py]: Why do you use `assert_allclose` instead of `assert_allequal`? Since we put the same numbers in the algorithms, there should be no numerical differences...

**Comment by mbillingr** [comment]
What about travis?
There are some changes in this PR that I think do not quite fit the topic. Moving functions between `utils` and `datatools` is a good idea, but not really related to the random state?

**Comment by cle1109** [comment]
> What about travis?

It's working now. I guess it was related to a `conda` update, which rendered our `numpy` version non-functional. Upgrading `numpy` to 1.8.2 solved the problem.

However, I don't understand why Travis doesn't trigger Coveralls anymore. Any ideas?

> There are some changes in this PR that I think do not quite fit the topic. Moving functions between `utils` and `datatools` is a good idea, but not really related to the random state?

Yes, it was related. Since I added `check_random_state` to `utils` and `acm` required `atleast_3d` from `datatools`, this resulted in a cyclic import. Moving `acm` solved the issue, but of course I could have also moved the `import` to the bottom of the file (which is ugly).

File: scot/tests/test_ooapi.py (hunk 3 above)
Comment (cle1109) [inline scot/tests/test_ooapi.py]: You're right, I'll replace it with `assert_allequal`.

File: scot/csp.py (hunk 1 above)
Comment (mbillingr) [inline scot/csp.py]: Hehe, ok, but now we have to pollute this PR by discussing why this is necessary.

Why is it necessary? :)

File: scot/datatools.py (hunk 2 above)
Comment (mbillingr) [inline scot/datatools.py]: Oh, I missed that.

File: scot/csp.py (hunk 1 above)
Comment (cle1109) [inline scot/csp.py]: Because we're dealing with Hermitian matrices. If our matrices are not Hermitian, this indicates a problem with our covariance matrices (e.g. due to a lack of data and/or no regularization). In such cases, `eigh` fails and tells us that something is wrong. `eig` on the other hand silently produces complex eigenvalues, which we usually ignored. For example, there was a problem with one test case, which generated a sine as input signal.

**Comment by mbillingr** [comment]
> However, I don't understand why Travis doesn't trigger Coveralls anymore. Any ideas?

The builds show up on the coveralls page, so I guess the problem is related to coveralls, not to travis. Otherwise I have no idea, yet.

> cyclic import.

I understand.

**Comment by mbillingr** [comment]
Btw, 
> Why did coverage increase? What is the reference/baseline?

The baseline is the current master. At the bottom of the report (coveralls page) it links to the baseline commit.

**Comment by cbrnr** [comment]
@mbillingr can you check if we've caught all randomness? And if so, please feel free to merge.

**Comment by mbillingr** [comment]
Can you rebase?

**Comment by cbrnr** [comment]
Yup. Done.
